framework/calico_framework.py

```diff
@@ -0,0 +1,673 @@
+        self.unreserve_offers_by(test)
+
+        # If there's no other tests, we're done here
+        if self.all_tests_complete():
+            self.report_results_and_exit()
+
+    def registered(self, driver, frameworkId, masterInfo):
+        """
+        Callback used when the framework is succesfully registered.
+        """
+        _log.info("REGISTERED: with framework ID %s", frameworkId.value)
+
+    def resourceOffers(self, driver, offers):
+        """
+        Triggered when the framework is offered resources by mesos.
+        """
+        unreserved_offers_by_slave_id = {}
+
+        # Send new offers to tasks which reserved them, or add to unreserved
+        # offers pool
+        for offer in offers:
+            new_offer = Offer(offer)
+            if new_offer.slave_id in self.test_by_slave_id:
+                _log.info("New offer %s Sent to reserved slave", new_offer)
+                self.test_by_slave_id[new_offer.slave_id].launch({new_offer.slave_id: new_offer})
+            else:
+                _log.info("New offer %s Moved to unreserved pool", new_offer)
+                unreserved_offers_by_slave_id[new_offer.slave_id] = new_offer
+
+        # Offer unreserved offers to unstarted tests
+        if unreserved_offers_by_slave_id:
+            _log.info("Offering %d unreserved offers to Unstarted Tests", len(unreserved_offers_by_slave_id))
+            for test in self.tests:
+                if test.running() or test.complete():
+                    continue
+
+                _log.info("\t%s", test)
+                # Offer to the test
+                reserved_offers_by_slave_id = test.can_run_on(unreserved_offers_by_slave_id)
+                if reserved_offers_by_slave_id:
+                    # Save slave reservations for the future
+                    for slave_id in reserved_offers_by_slave_id:
+                        self.test_by_slave_id[slave_id] = test
+
+                    # Send reserved offers to slave
+                    offers_for_test = {}
+                    for slave_id, offer in reserved_offers_by_slave_id.iteritems():
+                        offers_for_test[slave_id] = offer
+                        # Remove offer from unreserved pool
+                        del(unreserved_offers_by_slave_id[slave_id])
+                    test.launch(offers_for_test)
+
+        for offer in unreserved_offers_by_slave_id.values():
+            _log.debug("Declining unused offer: %s" % offer)
+            driver.declineOffer(offer.offer.id)
+
+        # Manage timeouts. First, Search to see if any tests are running
+        test_is_running = False
+        for test in self.tests:
+            if test.running():
+                test_is_running = True
+                break
+
+        # No running tests? Kill timed out tests and start timers
+        if not test_is_running:
+            for test in self.tests:
+                if not test.running() and not test.complete() and \
```

Comment:
Ok, so tests can be running or complete.  It seems intuitive that they cannot be both at the same time.  Could we then introduce a new test state: `not_started`?  Does that cover all the states a test can be in?  Then you could centralize `.running()` and `.complete()` with a single `.test_state()` that returns one of 3 named constants.